### ISSUE TITLE
Third party extensions

### ISSUE THREAD

**Issue by blink1073** [jupyter/jupyterlab#224]
JupyterLab will hold the base-level Webpack config for its externals library configuration and a script used to generate the externals for a given third party extension.

The script recursively concats the externals of the upstream jupyter lab extensions, where an extension is a dependency that has a valid ['jupyter']['lab'] config in its package.json.

A third party extension specifies its entry point in ['jupyter']['lab']['main'] and may optionally provide a custom externals configuration in ['jupyter']['lab']['externals'] if something other than the default behavior is desired.

In the cookecutter template, we provide a path to bridge lab and notebook by using the ['jupyter']['notebook'] config, which will assume an all-amd environment.

We propose using labextension to differentiate from nbextension.

The default stance for a third party library is an AMD bundle that exports its API.
Externals synchronously imported by JupyterLab itself will use script tags with accompanying SystemJS config.
Asynchronous imports in JupyterLab will use SystemJS.

The server can use the semver Python library to validate all of the extensions while populating the html template.

Open questions:

How to enable third party extensions to override a default plugin?

cf https://github.com/bollwyvl/widget-cookiecutter/tree/jupyterlab-extension for what a consumer might look like (still a wip).

**Comment by blink1073** [comment]
I verified this works by consuming a `jupyter-js-services` bundle created this way from the `web2` example in `ipywidgets`, and I got the appropriated objects from `jupyter-js-services`.

**Comment by blink1073** [comment]
cc @SylvainCorlay @jasongrout

**Comment by blink1073** [comment]
cf https://github.com/jupyter/jupyter-js-services/pull/162

**Comment by blink1073** [comment]
Here's what I did (after copying the `jupyter-js-services` bundle to the folder):

```diff
diff --git a/jupyter-js-widgets/examples/web/index.js b/jupyter-js-widgets/examples/web/index.js
--- a/jupyter-js-widgets/examples/web/index.html
+++ b/jupyter-js-widgets/examples/web/index.html
@@ -1,5 +1,6 @@
 <html>
     <head>
+        <script src="jupyter-js-services.js"></script>
         <script src="index.built.js"></script>
     </head>
     <body />
index 616f3e4..5039983 100644
--- a/jupyter-js-widgets/examples/web/index.js
+++ b/jupyter-js-widgets/examples/web/index.js
@@ -2,6 +2,8 @@
 // Create a widget manager instance.
 var WidgetManager = require('./manager').WidgetManager;
 var uuid = require("jupyter-js-widgets").uuid;
+var services = require('jupyter-js-services');
+console.log(services);

 document.addEventListener("DOMContentLoaded", function(event) {
     var manager = new WidgetManager(document.body);
diff --git a/jupyter-js-widgets/examples/web/webpack.config.js b/jupyter-js-widgets/examples/web/webpack.config.js
index 94b1b74..f29fd91 100644
--- a/jupyter-js-widgets/examples/web/webpack.config.js
+++ b/jupyter-js-widgets/examples/web/webpack.config.js
@@ -1,4 +1,7 @@
 module.exports = {
+  output: {
+    libraryTarget: 'umd'
+  },
   module: {
     loaders: [
       { test: /\.css$/, loader: "style-loader!css-loader" },
@@ -7,4 +10,7 @@ module.exports = {
       { test: /\.(jpg|png|gif)$/, loader: "file" },
     ]
   },
+  externals: {
+    'jupyter-js-services': 'umd jupyter-js-services'
+  }
 }
```

**Comment by blink1073** [comment]
<img width="636" alt="screen shot 2016-07-02 at 10 04 51 am" src="https://cloud.githubusercontent.com/assets/2096628/16540956/75de32ce-403c-11e6-9ea9-16ed909eec8f.png">

**Comment by blink1073** [comment]
![bigyes](https://cloud.githubusercontent.com/assets/2096628/16540970/d58e6644-403c-11e6-9e77-84c0992bfa05.gif)

**Comment by blink1073** [comment]
More thoughts:
Define the public api - all of phosphor and jupyter-services, and an api file in jupyterlab

Third party extensions make sure to exclude these from their published webpack bundle. 

We can require their bundle at runtime.

**Comment by jasongrout** [comment]
This is blocked on the monorepo, since it will depend on packaging up lots of the phosphor things.

**Comment by blink1073** [comment]
I found this, which we can use to assemble the pieces: https://www.npmjs.com/package/webpack-system-register

**Comment by blink1073** [comment]
This allows us to use `System.import` instead of global variables.

**Comment by blink1073** [comment]
Theoretically, we could use `System.bundle` to do the whole thing, but we'd be trading away all the work we've done so far in configuring `WebPack` loaders and bundles.

**Comment by jasongrout** [comment]
I think we'll want to switch away from global variables anyway. I'll talk
to you about it in Monday, but basically we should be able to load
dependencies like underscore and backbone. Not sure if that works
currently...

**Comment by blink1073** [comment]
We don't have a way to provide things like underscore yet.  My thinking is that third party extensions could provide their own dependencies if they are not already available from upstream.  Basically, we'd be providing a federated npm that is dynamically loaded.

**Comment by blink1073** [comment]
Proposed design goals:

- Automated and simple to reason about for third party extension authors
- Robust
- Use semver
- No global variables
- Load only one module of each version
- Support multiple versions if necessary
- Load from multiple directories
- Support CSS and JSON
- Support resolving modules dynamically (e.g. comm targets, CodeMirror modes/addons)

**Comment by blink1073** [comment]
More thoughts:

- Stick with Webpack for node resolution and loaders

- We need everything available synchronously to WebPack

- So, keep the `jupyter.externals` namespace

- Add version mangling to the externals (e.g.) `jupyter.externals['phosphor^0.5.0']`

- If downstream depends on an incompatible version of a lib, that version gets included in its bundle.

- We explicitly list the modules we provide using `find_imports`

- This is what we use to build each shim

**Comment by blink1073** [comment]
Actually, [webpack-system-register](https://github.com/CanopyTax/webpack-system-register#configuration-options) will take care of making things synchronously available to each bundle, so we can use `SystemJS` for the infrastructure instead of the global var.

**Comment by blink1073** [comment]
A recap of a conversation we just had on gitter.

- Use Webpack and AMD for everything, rather than fighting too much between different module loaders.
- Add a shim that re-exports internal modules using `define` or a `require` config. 
- We shim all of the modules that we explicitly use and all of our own internal modules.
- Each extension adds itself and any of its imports that are not already provided.
- If an extension depends on a conflicting version of a library, we warn loudly and do not export those modules.
- This could be used in `jupyter/notebook`, since it results in bundles that have their individual paths exported using AMD

**Comment by jasongrout** [comment]
> If an extension depends on a conflicting version of a library, we warn loudly and do not export those modules.

If we have two plugins independently providing, say, d3, this won't help.

It'd be nice if there was an AMD semver require, where I could require a semver of a particular module, and it would resolve to the relevant version of the module. At load time, *every* version provided would be available, and on require, it would resolve to the newest one that satisfied the semver. I think you mentioned something like this at some point.

**Comment by blink1073** [comment]
As in, jspm :smile:?

**Comment by jasongrout** [comment]
Uh, I guess? Nice!

**Comment by jasongrout** [comment]
More notes from conversations with @blink1073. We have externals and shims - shims are things we are providing, and externals are the way to resolve to what we are providing.

* Explicitly list things we are providing (i.e., things shims will be built for). Generally list the things we think dependencies will need to share and not have their private copies. For the ipywidgets, provide jupyter-js-widgets, and maybe backbone and underscore? Each provided library generates a separate webpack bundle. This gets a bit tricky because the webpack configs for each library you are providing needs to manually state what dependency externals will be provided by the plugin itself. (so the widgets lab extension will have to configure jupyter-js-widgets to rely on an external of backbone and underscore...). The bundle includes the library name and version. The library bundle includes all accessible modules in the library. The bundle also has to include a manifest giving what externals it is assuming (and the appropriate semvers for those externals).
* Shims are built to load into a global package registry (`jupyter.externals`) the package name, path, and version number for each module in the external library.
* The externals webpack helper walks the ancestor tree and mangle imports to include a *semver* version of the library (pulled from package.json). The resolver calls a runtime function with the library name, path, and semver, and the runtime function resolves to a loaded package in `jupyter.externals`. E.g., `jupyter.resolve('phosphor/lib/ui/widget', '^0.6.0')`

* Loading bundles on a page: webpack assumes things are provided synchronously. Loading using script tags means the server has to resolve the dependency tree (with semver). Each plugin package has an extension manifest file that lists each external bundle it provides, and the externals dependencies for that bundle (with semver). This manifest file ...

* Each npm package can provide multiple external bundles. Each npm package provides at most one plugin. The plugin manifest lists for each bundle:
  - name (an npm package name)
  - version number
  - list of externals it assumes (with their semvers)
TODO: some way of saying what to load to get the IPlugin data structure.

Initial boostrap: server templates a bunch of script tags to load bundles in order. Gets each IPlugin on the page somehow (maybe through requirejs?). Then the application is loaded and given the list of IPlugins, and then we call `application.run()`

Externals we are providing - if the library has a package.json `main` entry point, that is what we export. Otherwise we provide all .js and .css files in the package.json `directory.lib` directory.

**Comment by blink1073** [comment]
FYI There is a plugin for RequireJS that allows you to return a promise in the `define()` block.

https://github.com/jokeyrhyme/requirejs-promise

Great example here:
http://stackoverflow.com/a/26445338

**Comment by blink1073** [comment]
I played with the WebPack DLLPlugin:

Pros:
- Does not execute any code
- Auto-creates shims for all modules
- Modules are available synchronously

Cons:
- The dll needs to export a global variable
- We would need to figure out how to dynamically use the manifest(s)

My current plan is to not export anything explicitly from the bundles, but have them only `define()` each module using a shim that we create.
The WebPack externals would mangle to `amd foo`, and use these defined modules.

**Comment by blink1073** [comment]
Closing in favor of #728.
